### PR TITLE
quantlib 1.7

### PR DESCRIPTION
**Pull request by enricodetoma**
QuantLib 1.6.1 doesn't compile correctly with Boost 1.59, so this is a necessary update.

### CODE REVIEW

**Comment by DomT4** [comment]
Apparently doesn't compile here either:
```
ld: library not found for -lboost_thread
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [FRA] Error 1
make[1]: *** [install-recursive] Error 1
make: *** [install-recursive] Error 1
```

I have a `-lboost_thread-mt` locally but no plain `-lboost_thread`

**Comment by enricodetoma** [comment]
I had to do this (in /usr/local/lib):
```
ln -s libboost_thread-mt.a libboost_thread.a
ln -s libboost_thread-mt.so libboost_thread.so
```
to be able to link successfully with boost_thread.

But this is not the bug I mentioned.
The bug I cited for 1.6.1 was a compile time bug (see this thread for reference: http://quantlib.10058.n7.nabble.com/ambiguous-symbol-error-building-Quantlib-1-6-with-Boost-1-59-unit-test-framework-td16917.html)

About this link bug with boost_thread, it seems to be boost related rather than quantlib related (see: http://stackoverflow.com/questions/20974036/libboost-tread-missing-on-os-x-boost-from-homebrew and: http://stackoverflow.com/questions/12518973/link-with-boost-homebrew-mac-c).

**Comment by DomT4** [comment]
You might be able to give this a nudge in the right direction by mimicking something similar to [this](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/povray.rb#L57) here.

**Comment by enricodetoma** [comment]
Thank you for your suggestion.

Now doing tihs:
```
  def install
    # include the boost thread library to resolve compilation conflicts
    ENV["LIBS"] = "-lboost_thread-mt"
```

I still can't get rid of the -lboost_thread dependency (-lboost_thread-mt simply gets added):
```
==> make install
Last 15 lines from /Users/enrico/Library/Logs/Homebrew/quantlib/02.make:
make[3]: Nothing to be done for `install-data-am'.
Making install in FittedBondCurve
clang++ -DHAVE_CONFIG_H -I. -I../../ql  -I../.. -I../..   -g -O2 -c -o FittedBondCurve.o FittedBondCurve.cpp
/bin/sh ../../libtool  --tag=CXX   --mode=link clang++  -g -O2   -o FittedBondCurve FittedBondCurve.o ../../ql/libQuantLib.la  -lboost_thread-mt
libtool: link: clang++ -g -O2 -o .libs/FittedBondCurve FittedBondCurve.o  ../../ql/.libs/libQuantLib.dylib -lboost_thread-mt
make[3]: Nothing to be done for `install-data-am'.
Making install in FRA
clang++ -DHAVE_CONFIG_H -I. -I../../ql  -I../.. -I../..   -g -O2 -c -o FRA.o FRA.cpp
/bin/sh ../../libtool  --tag=CXX   --mode=link clang++  -g -O2   -o FRA FRA.o ../../ql/libQuantLib.la -lboost_thread  -lboost_thread-mt
libtool: link: clang++ -g -O2 -o .libs/FRA FRA.o  ../../ql/.libs/libQuantLib.dylib -lboost_thread -lboost_thread-mt
ld: library not found for -lboost_thread
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [FRA] Error 1
make[1]: *** [install-recursive] Error 1
make: *** [install-recursive] Error 1
```

Do you have any suggestion how to remove the -lboost_thread dependency?

**Comment by enricodetoma** [comment]
The configure file for QuantLib contains these lines:

```
 boost_thread_lib="-lboost_thread -lboost_system -lpthread"
 LIBS="$ql_original_LIBS $boost_thread_lib" 
```

So the dependency on boost_thread is hard encoded, and can't be easily converted to using boost_thread-mt.

Out of curiosity: why does the boost formula NOT produce a libboost_thread.a symlink to libboost_thread-mt.a to be compatible with any other library that needs -lboost_thread?

**Comment by mikemcquaid** [comment]
> Out of curiosity: why does the boost formula NOT produce a libboost_thread.a symlink to libboost_thread-mt.a to be compatible with any other library that needs -lboost_thread?

This shouldn't be necessary; other libraries detect the correct version. I'd think that the `./configure` file in this case needs patched (and the patch submitted upstream) to check for which version is available.

**Comment by enricodetoma** [comment]
@mikemcquaid you are right: the QuantLib maintainers introduced a bug in one of the subprojects by hard encoding the -lboost_thread flag in a Makefile (when otherwise detecting the correct version automatically).
This bug is present only in the latest 1.7 version, and they will fix it in the upcoming 1.8 release.

So, for now, I provided a patch as a Github gist, and I added a new commit to this pull request which uses the patch. Now it should compile and link without errors (I tested it with the -c++11 flag in my environment).

Please review it.

To submit the patch upstream, shall I need to open a pull request in the patch repository, then update the url here when you accept it there?

In any case, when they release QuantLib 1.8, this patch will not be necessary any more.

File: Library/Formula/quantlib.rb [hunk 1]
```diff
@@ -22,6 +22,13 @@ class Quantlib < Formula
 
   option :cxx11
 
+  # fix for quantlib 1.7 linking conflicts with the boost thread library
+  # this patch must be removed when quantlib 1.8 is released, because quantlib maintainers fixed the bug
```
Comment (mikemcquaid) [inline Library/Formula/quantlib.rb]: Can you link to the upstream commit/discussion verifying this bug is fixed as a comment in the Ruby file? Thanks!

**Comment by mikemcquaid** [comment]
> To submit the patch upstream, shall I need to open a pull request in the patch repository, then update the url here when you accept it there?

Submit it upstream generally means to the creators of the software i.e. quantlib.

If you could also do that it'd be great, thanks :+1: .

**Comment by enricodetoma** [comment]
They already fixed it, for the next release, I found the patch in one of their commits.

For Homebrew, you can choose to accept my current formula with the patch here (which works) or just wait until they release next version of QuantLib, where an updated formula will need no patch.

File: Library/Formula/quantlib.rb (hunk 1 above)
Comment (enricodetoma) [inline Library/Formula/quantlib.rb]: Here it is: https://github.com/lballabio/QuantLib/commit/d1909593d9f36c6703966460fb48773792facd7e

Comment (mikemcquaid) [inline Library/Formula/quantlib.rb]: Please link to it as a comment in the formula Ruby file. Thanks!

Comment (enricodetoma) [inline Library/Formula/quantlib.rb]: Done

**Comment by mikemcquaid** [comment]
Looks good to me. :+1: once green.

**Comment by enricodetoma** [comment]
Travis CI: The job exceeded the maxmimum time limit for jobs, and has been terminated.

QuantLib build is very long...

**Comment by DomT4** [comment]
Merged in https://github.com/Homebrew/homebrew/commit/2325e73fea7c7941de4ca2e8b71777af954eeac8. Thank you for your contribution to Homebrew @enricodetoma; we appreciate it! :smiley_cat:

**Pull request closed without merging**